release-notes/swan-lake-beta4-release-note.md

```diff
@@ -34,9 +34,479 @@ If you have not installed Ballerina, then download the [installers](/downloads/#
+   int? g = a << c;
+   int? h = a & b;
+   int? i = -a;
+   int? j = +c; // result is `()`
+}
+```
+
 #### Improvements
 
-#### Bug Fixes
+##### Restrictions When Calling a Function or a Method in a Match Guard
+
+Restrictions have been introduced for when a function or a method is called in a match guard, in order to ensure that the match guard does not mutate the value being matched.
+
+A function or method call is now allowed in a match guard only if it meets one of the following conditions.
+- the type of the expression following `match` is a subtype of `readonly` or
+- the function/method is `isolated` and the types of any and all arguments are subtypes of `readonly`
+
+The following will now result in compilation errors.
+
+```ballerina
+type Data record {
+   string name;
+   boolean valid;
+   int id?;
+   decimal price?;
+};
+ 
+public function main() {
+   Data data = {name: "Jo", valid: false};
+   [int, decimal] currentValues = [1234, 20.5];
+ 
+   match data {
+       var {id, price} => {
+       }
+       // Now results in compilation errors for the match guard since neither the type of the matched
+       // expression nor the types of the arguments are subtypes of `readonly`.
+       var {name} if stillValid(data, currentValues) =>  {
+       }
+   }
+}
+ 
+isolated function stillValid(Data data, [int, decimal] values) returns boolean {
+   // ...
+   data.id = values[0];
+   data.price = values[1];
+   return false;
+}
+```
+
+##### Improved support for unreachability
+
+Unreachability analysis has been improved for if-else statements and `while` statements. Constant conditions that are known to be either true or false at compile-time are now considered in unreachability analysis.
+
+The following conditions are taken into consideration in the analysis of unreachability.
+
+1. If a statement block is unreachable, then every statement in it is unreachable.
+2. `if` statements with constant conditions are not errors except insofar as they lead to statements being unreachable.
+3. An `is` expression is constant `true` if the static type of the expression is a subtype of the type against which the check is done. 
+4. Calling a function with a return type of `never` cannot complete normally, making subsequent code unreachable.
+
+```ballerina
+function fn1() {
+    if false {
+        int a = 10; // this will now result in a compilation error: unreachable code
+    }
+
+    while false {
+        int a = 10; // this will now result in a compilation error: unreachable code
+    }
+}
+
+function fn2() {
+    if true {
+        int a = 10;
+    } else {
+        int a = 20; // this will now result in a compilation error: unreachable code
+    }
+}
+
+function fn3() {
+    if true {
+        return;
+    }
+    int a = 10; // this will now result in a compilation error: unreachable code
+}
+
+function fn4() {
+    while true {
+        return;
+    }
+    int a = 10; // this will now result in a compilation error: unreachable code
```

Comment:
@lasinicl, should we replace the `int a = 10;` lines with some `io:println` calls maybe? Since these result in a bunch of "unused variable" warnings also now.

Reply:
Fixed in https://github.com/ballerina-platform/ballerina-release/pull/1346

Reply:
Thanks, merged.